### PR TITLE
puffer_tetris

### PR DESCRIPTION
**Pull request by Hadrien-Cr**
I implemented Tetris, following the specifications and the NN architecture from jumanji (https://instadeepai.github.io/jumanji/environments/tetris/). NN architecture is written in `ocean/torch.py`

Features:
- Same action space as jumanji: The agent chooses a target column and a target rotation, and the piece falls instantly on the chosen spot. Thus the action space is (N_cols x 4)
- An action mask is calculated at each step and used to mask the invalid actions on the NN outputs.
- Same reward as jumanji: the reward is based on the number of lines cleared and the reward_list [0, 40, 100, 300, 1200]. If no lines are cleared, the reward is 0. 
- A timeout of 400 steps is used.
- The Agent can observe n of the next pieces to come (the parameter is called `deck_size`)
- The Agent observes the current grid, and process the.

On my machine the Forward/Learn take 100% of the load, so I have a big GPU bottleneck. Anyway, you can expect 2000 score in 200M-300M steps, which falls a little bit behind the benchmark from jumanji paper https://arxiv.org/abs/2306.09884. Obviously the SPS are better with pufferlib :)

### CODE REVIEW

File: pufferlib/ocean/tetris/tetris.h [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+    int preview_rows;
+    int preview_target_rotation;
+    int preview_target_col;
+} Client;
+
+typedef struct Tetris {
+    Client* client;
+    Log log;
+    float* observations;
+    float* actions;
```
Comment (jsuarez5341) [inline pufferlib/ocean/tetris/tetris.h]: Change to int! Will not train as is because Python is writing the bits of an int to a float

File: pufferlib/ocean/tetris/tetris.h [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+void update_deck(Tetris* env) {
+    env->tetromino_deck[env->current_position_in_deck] = rand() % NUM_TETROMINOES;
+    env->current_position_in_deck = (env->current_position_in_deck +1) % env->deck_size;
+    env->current_tetromino = env->tetromino_deck[env->current_position_in_deck];
+}
+
+bool can_place_tetromino(Tetris* env, int tetromino_idx, int target_col, int target_rotation, int target_row) {
+    for (int r = 0; r < SIZE; r++) {
+        for (int c = 0; c < SIZE; c++) {
+            if (TETROMINOES[tetromino_idx][target_rotation][r][c] == 1) {
```
Comment (jsuarez5341) [inline pufferlib/ocean/tetris/tetris.h]: invert conditionals where trivial

File: pufferlib/ocean/tetris/tetris.h [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+    env->terminals[0] = 0;
+    env->rewards[0] = 0.0;
+    env->step += 1;
+    int action = env->actions[0];
+    int col = action%env->n_cols;
+    int rotation = action/env->n_cols;
+    int lines_deleted = 0;
+    int landing_row;
+
+    if (env->action_mask[action]) {
```
Comment (jsuarez5341) [inline pufferlib/ocean/tetris/tetris.h]: Waste to compute the whole atn mask if you don't need it. We don't have train-level action mask supporting right now. It's very effective in specific problems, but not generally viable because computing action mask a-priori can be slow or intractable

File: pufferlib/ocean/torch.py [hunk 4]
```diff
@@ -861,3 +861,85 @@ def decode_actions(self, flat_hidden):
...
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.cnn_channels =  cnn_channels   
+
+        self.n_cols = env.n_cols
+        self.n_rows = env.n_rows
+        self.scalar_input_size = (1 + 7 * env.deck_size)
+        self.is_continuous = isinstance(env.single_action_space, pufferlib.spaces.Box)
+
+        self.conv_grid = nn.Sequential(
```
Comment (jsuarez5341) [inline pufferlib/ocean/torch.py:883]: Too chonky, guarantee 2 layers solves

File: pufferlib/ocean/torch.py [hunk 5]
```diff
@@ -861,3 +861,85 @@ def decode_actions(self, flat_hidden):
...
+    def encode_observations(self, observations, state=None):
+        grid_info = observations[:,0:(self.n_cols * self.n_rows)].view(-1, 1, self.n_rows, self.n_cols).float() # (1, n_rows,n_cols)
+        scalar_info = observations[:, (self.n_cols * self.n_rows):(self.n_cols * self.n_rows + self.scalar_input_size)].float()
+
+        grid_feat = self.conv_grid(grid_info) # (I, 2, n_cols,)
+        scalar_feat = self.fc_scalar(scalar_info) # (I,)
+        
+        grid_feat = grid_feat.permute(0,3,2,1) # (n_cols, 2, I,)
+        grid_feat = grid_feat.reshape(-1, grid_feat.shape[1], grid_feat.shape[2] * grid_feat.shape[3]) # (n_cols, 2 * I,)
+        scalar_feat = scalar_feat.unsqueeze(1).repeat(1, self.n_cols, 1)  # (N, self.n_cols, I)
```
Comment (jsuarez5341) [inline pufferlib/ocean/torch.py]: Repeat seems sketch. Should just be an unsqueeze

**Review by jsuarez5341 — CHANGES_REQUESTED** [review]
Fix buffer type + rewards, try a 2 layer policy, train a couple things, and PR - will help finetune from there!

**Comment by Hadrien-Cr** [comment]
https://github.com/user-attachments/assets/8acf863b-9eba-4a51-89d9-1f9f87879276

**Comment by Hadrien-Cr** [comment]
@jsuarez5341 I refactored Tetris to a real time game! Training went all good, reaching 11k score in 300M steps whereas my PB sits a little bit over 12k.

Only weird thing is that AI does not see to care about "rotate" action, new meta?

**Comment by ClashLuke** [comment]
Excellent work, thank you for pushing this!
